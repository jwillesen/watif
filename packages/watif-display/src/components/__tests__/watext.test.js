import React from 'react'
import {render, fireEvent} from 'react-testing-library'
import Watext from '../watext'

it('creates a paragraph for a top level text', () => {
  const {getByText} = render(<Watext watext={<text>foo</text>} />)
  getByText('foo')
})

it('creates a paragraph for a top level pg', () => {
  const {getByText} = render(<Watext watext={<pg>foo</pg>} />)
  getByText('foo')
})

it('creates multiple paragraphs inside text', () => {
  const {getByText} = render(
    <Watext
      watext={
        <text>
          <pg>foo</pg>
          <pg>bar</pg>
        </text>
      }
    />
  )
  getByText('foo')
  getByText('bar')
})

it('creates multiple paragraphs inside pg', () => {
  const {getByText} = render(
    <Watext
      watext={
        <pg>
          <pg>foo</pg>
          <pg>bar</pg>
        </pg>
      }
    />
  )
  getByText('foo')
  getByText('bar')
})

it('creates paragraphs using paragraph tag as a divider', () => {
  const {getByText} = render(
    <Watext
      watext={
        <text>
          foo
          <pg />
          bar
        </text>
      }
    />
  )
  getByText('foo')
  getByText('bar')
})

it('creates item buttons', () => {
  const {getByText} = render(
    <Watext
      watext={
        <text>
          here is <item id="an-item">an item</item> for you
        </text>
      }
    />
  )
  const itemButton = getByText('an item')
  expect(itemButton.tagName).toBe('BUTTON')
})

it('alerts if the item has child elements', () => {
  const {getByText} = render(
    <Watext
      watext={
        <text>
          here is{' '}
          <item id="some-item">
            a special <pg>type of thing</pg>
          </item>{' '}
          for you
        </text>
      }
    />
  )
  getByText(/alert/i)
})

it('alerts if the item lacks an id', () => {
  const {getByText} = render(
    <Watext
      watext={
        <text>
          <item>an item</item>
        </text>
      }
    />
  )
  getByText(/alert/i)
})

it('alerts if it tag is not recognized', () => {
  const {getByText} = render(
    <Watext
      watext={
        <text>
          stuff <somethingbad>bad things</somethingbad> here
        </text>
      }
    />
  )
  getByText(/alert/i)
})

it('invokes the callback when an item is clicked', () => {
  const onClick = jest.fn()
  const {getByText} = render(
    <Watext
      watext={
        <text>
          here is <item id="an-item">an item</item> to click
        </text>
      }
      onItemClick={onClick}
    />
  )
  fireEvent.click(getByText('an item'))
  expect(onClick).toHaveBeenCalledWith('an-item')
})
