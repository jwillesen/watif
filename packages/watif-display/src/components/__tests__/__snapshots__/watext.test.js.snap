// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alerts if it tag is not recognized 1`] = `
<div
  className="content"
>
  <div
    className="paragraph"
    key="0,0"
  >
    stuff 
  </div>
  <div
    className="paragraph"
    key="0,1"
  >
    <Alert
      variant="error"
    >
      unrecognized watext element: 
      somethingbad
    </Alert>
  </div>
  <div
    className="paragraph"
    key="0,2"
  >
     here
  </div>
</div>
`;

exports[`alerts if the item has child elements 1`] = `
<div
  className="content"
>
  <div
    className="paragraph"
    key="0,0"
  >
    here is 
  </div>
  <div
    className="paragraph"
    key="0,1"
  >
    <Alert
      variant="error"
    >
      Item elements may not contain other elements.
    </Alert>
  </div>
  <div
    className="paragraph"
    key="0,2"
  >
     for you
  </div>
</div>
`;

exports[`alerts if the item lacks an id 1`] = `
<div
  className="content"
>
  <div
    className="paragraph"
    key="0,0"
  >
    <Alert
      variant="error"
    >
      Item elements require an id property
    </Alert>
  </div>
</div>
`;

exports[`creates a paragraph for a top level paragraph tag 1`] = `
<div
  className="content"
>
  <div
    className="paragraph"
    key="0,0"
  >
    foo
  </div>
</div>
`;

exports[`creates and flattens multiple paragraphs when they are nested 1`] = `
<div
  className="content"
>
  <div
    className="paragraph"
    key="0,0"
  >
    foo
  </div>
  <div
    className="paragraph"
    key="0,1"
  >
    bar
  </div>
  <div
    className="paragraph"
    key="0,2"
  >
    baz
  </div>
</div>
`;

exports[`creates at least one paragraph for top level text 1`] = `
<div
  className="content"
>
  <div
    className="paragraph"
    key="0,0"
  >
    foo
  </div>
</div>
`;

exports[`creates item buttons 1`] = `
<div
  className="content"
>
  <div
    className="paragraph"
    key="0,0"
  >
    here is 
    <button
      className="item-button"
      key="an-item"
      onClick={[Function]}
    >
      an item
    </button>
     for you
  </div>
</div>
`;

exports[`creates multiple paragraphs with top level pg 1`] = `
<div
  className="content"
>
  <div
    className="paragraph"
    key="0,0"
  >
    foo
  </div>
  <div
    className="paragraph"
    key="0,1"
  >
    bar
  </div>
</div>
`;

exports[`creates multiple paragraphs with top level text 1`] = `
<div
  className="content"
>
  <div
    className="paragraph"
    key="0,0"
  >
    foo
  </div>
  <div
    className="paragraph"
    key="0,1"
  >
    bar
  </div>
</div>
`;

exports[`creates paragraphs using paragraph tag as a divider 1`] = `
<div
  className="content"
>
  <div
    className="paragraph"
    key="0,0"
  >
    foo
  </div>
  <div
    className="paragraph"
    key="0,1"
  >
    bar
  </div>
</div>
`;

exports[`renders provided text if watext is missing 1`] = `
<div
  className="content"
/>
`;
